update goss verify.yml to allow to pass host to run on

loomsen opened the pull request:
This PR allows for the goss verifier to run different tests on certain hosts.
Example: have a side-effect playbook, which makes keepalived move an IP to a different host in an active-active scenario. The verifier should verify that the IP actually moved to a different host, and that the side-effect playbook killed the supervised process.

Basically what you can do with testinfra by specifying the host in .get_hosts

```
testinfra_hosts = testinfra.utils.ansible_runner.AnsibleRunner(
    os.environ['MOLECULE_INVENTORY_FILE'])\
    .get_hosts('instance1')
```

You can do this for goss by prepending the test file with test_host_$hostname_as_given_in_molecule.yml.

#### PR Type

- Feature Pull Request

lwm commented on molecule/cookiecutter/scenario/verifier/goss/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/verify.yml:
```diff
@@ -10,37 +10,49 @@
+      command: "{{ goss_bin }} -g {{ item }} validate --format {{ goss_format }}"
       register: test_results
       with_items: "{{ test_files.stdout_lines }}"
+      ignore_errors: true
```
Should we really ignore here?

loomsen replied:
If we don't, and a test fails, we will never reach the next step to "Display details about the Goss results", will we?
Maybe a smarter failed_when condition could do it? I'll have to read up on goss' error codes then.

lwm replied:
Yep, makes sense!

lwm commented on molecule/cookiecutter/scenario/verifier/goss/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/verify.yml:
```diff
@@ -10,37 +10,49 @@
+      delegate_to: localhost
+      register: test_files
+
+    - name: debug
```
Should only be turned on if we're passing in debug arguments?

loomsen replied:
Isn't that what verbosity does? Or do you mean molecule debug arg? 
```
Note
Molecule will remove any options matching ‘^[v]+$’, and pass -vvv to the underlying ansible-galaxy command when executing molecule –debug.
```
So if you pass --debug, verbosity=3 should match this behavior.

lwm replied:
Cool.

lwm commented on test/scenarios/verifier/molecule/goss/verify.yml:
```diff
@@ -22,12 +22,22 @@
+      delegate_to: localhost
+      register: test_files
+
+    - name: debug
```
Same as above ^^^.

loomsen replied:
Same as above ^^^ :)

lwm replied:
:+1:

loomsen commented:
Hi, thank you for reviewing my PR :)

I set up a stripped down, real world example of the roles we actually use in production. It's dumbed down to a default vhost returning 502, but it shows the purpose of this PR.

https://github.com/loomsen/ansible-nginx-ha-cluster-pr-example

Another scenario would be a master-slave databse setup, where you want to setup the replication with ansible, and have a side effect playbook run to make sure the failover is working. 
Without this PR, there is no way to test that, because there is only one common goss test file for all hosts.

I hope this enlightens the dark. Does it?

loomsen commented:
Politely asking travis for a rebuild, the failure seems kind of random to me.

lwm commented:
Sorry that you're PR is sitting so long here @loomsen. Thanks for the further example.

I've given some more time on this. If I understand it correctly, this PR solves two issues:

  * Fixes the SHA for the goss binary as reported in https://github.com/ansible/molecule/issues/1618.
  * Allows to specify hosts which goss will run on (via `test_host_MYHOST.yml` files).

So yes, this definitely feels like something we need then. This PR is missing documentation for this change that would definitely make it more visible for other users. Also, can me perhaps use the more readable [find](https://docs.ansible.com/ansible/latest/modules/find_module.html) module to replace the `shell: ...` pattern matching?

And finally, how to test this? It will be hard to have any sense of sureness without that. If there isn't something already available to help you write tests, please let us know and we better get to making that happen too!

lwm commented on molecule/cookiecutter/scenario/verifier/goss/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/verify.yml:
```diff
@@ -10,37 +10,51 @@
+        state: directory
+
+    - name: Find Goss tests on localhost
+      shell: "find {{ lookup('env', 'MOLECULE_VERIFIER_TEST_DIRECTORY') }} \\( -name 'test_*.yml' -and -not -name 'test_host_*.yml' \\) -or -name 'test_host_{{ ansible_hostname }}*.yml'"
```
To be clear (after my last comment), I mean this pattern matching.

https://docs.ansible.com/ansible/latest/modules/find_module.html

loomsen commented:
OK, so I added the requested changes, some documentation and I changed the test scenario to reflect the new feature. 
What do you think?

I pulled in latest master which oddly introduced some linting errors.

webknjaz commented on molecule/cookiecutter/scenario/verifier/goss/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/verify.yml:
```diff
@@ -10,37 +10,58 @@
-    goss_url: "https://github.com/aelsabbahy/goss/releases/download/{{ goss_version }}/goss-linux-{{ goss_arch }}"
-    goss_test_directory: /tmp
+    goss_bin: /usr/local/bin/goss
+    goss_sha256sum: 53dd1156ab66f2c4275fd847372e6329d895cfb2f0bcbec5f86c1c4df7236dde.
```
Is that trailing dot a typo?
```suggestion
    goss_sha256sum: 53dd1156ab66f2c4275fd847372e6329d895cfb2f0bcbec5f86c1c4df7236dde
```

loomsen replied:
Hi, thanks, I removed the dot. Although I think it was there when I updated that sha256sum. However, it's gone now :)

loomsen commented:
I don't know why the builds fail.
https://travis-ci.com/loomsen/molecule/builds/98478819
That's the build from my fork, which is fine (except the lint errors).

After fetching latest master into my develop branch, some jobs also start failing.

https://travis-ci.com/loomsen/molecule/builds/98581045

Hence I think the failures are unrelated to my changes.
What do you guys think?

lwm commented:
Added to agenda in https://github.com/ansible/community/issues/427. Please join if you have time!

lwm commented:
Thanks for your patience @loomsen. Our CI is back in shape. It reported:

https://travis-ci.com/ansible/molecule/jobs/179293976

```
Warning, treated as error:
/home/travis/build/ansible/molecule/molecule/verifier/goss.py:docstring of molecule.verifier.goss.Goss:51:Error in "code-block" directive:
maximum 1 argument(s) allowed, 18 supplied.
.. code-block::
    test_default.yml (will run on all hosts)
    test_host_instance1.yml (will run only on instance1)
    test_host_instance2.yml (will run only on instance2)
ERROR: InvocationError for command '/home/travis/build/ansible/molecule/.tox/doc/bin/python setup.py build_sphinx -n -W --builder=html' (exited with code 1)
```

decentral1se commented:
Just missing the `--signoff` DCO for the commit :cold_sweat:

loomsen commented:
Gah, every single time.

loomsen commented:
Woohoo, finally :)

decentral1se commented:
Oh no .... we still didn't merge this ....

@loomsen, can you resolve the final merge conflicts? (sorry :pray:).

We're waiting on one final issue wrt. packaging and then we can merge this (for real).

:sweat_smile:

loomsen commented:
OK, so I pulled in the latest changes. Code style check fails, which shouldn't be related to this PR, and also a test randomly fails at destroy stage, maybe one of you could re-run just this build, if possible? 

@decentral1se thanks for bearing with me :)

decentral1se commented:
No, thank you!!!

decentral1se commented:
It appears to be related to https://github.com/ansible/molecule/pull/1911!

decentral1se commented:
![](https://thumbs.gfycat.com/SeveralDenseAfricanbushviper-small.gif)
